wallet/src/wallet/mod.rs

```diff
@@ -187,19 +214,37 @@ impl<B: storage::Backend> Wallet<B> {
         self.accounts.keys()
     }
 
-    // TODO: this should not be public, as specified by BIP44 accounts must be created sequentially
-    // and a new next account should be rejected if the previously created one has no transactions
-    // associated with it
-    pub fn create_account(&mut self, account_index: U31) -> WalletResult<()> {
+    pub fn number_of_accounts(&self) -> usize {
+        self.accounts.len() + self.unsynced_accounts.len()
+    }
+
+    pub fn create_account(&mut self) -> WalletResult<U31> {
         ensure!(
-            !self.accounts.contains_key(&account_index),
-            WalletError::AccountAlreadyExists(account_index)
+            self.unsynced_accounts.is_empty(),
+            WalletError::LastAccountNotInSync
         );
 
+        let next_account_index = self.accounts.last_key_value().map_or(
+            Ok(U31::ZERO),
+            |(last_account_index, last_account)| {
+                if last_account.has_transactions() {
+                    last_account_index.plus_one().map_err(|_| {
+                        WalletError::AbsoluteMaxNumAccountsExceeded(
+                            last_account_index.into_u32() + 1,
```

Comment:
I think this is wrong. You want to show the maximum, not the next-to-maximum, right? This may or may not overflow based on whether the derivation is hard. Just use the last available value, without the +1, which is the absolute maximum.

Reply:
removed the +1